### PR TITLE
Add .spec.serviceAccount.mountToken config option

### PR DESCRIPTION
**Pull request by idlewis**
Deals with RCO portion of OpenLiberty/open-liberty-operator#462

### CODE REVIEW

**Review by leochr — COMMENTED** [review]
@idlewis Thanks for the PR. I'll suggest creating a new config section for Service Account to group relevant fields. With the addition of the mount option, we'll have 2, but any config related to the service account can be added here in the future:

```
spec:
   serviceAccount:
      name:
      mountToken:
```

The existing field `.spec.serviceAccountName` should be marked as deprecated. But we should look it up if `.spec.serviceAccount.name` is not set.

File: api/v1/runtimecomponent_types.go [hunk 1]
```diff
@@ -144,6 +148,16 @@ type RuntimeComponentSpec struct {
 	SecurityContext *corev1.SecurityContext `json:"securityContext,omitempty"`
 }
```
Comment (leochr) [inline api/v1/runtimecomponent_types.go:150]: Description is needed here. It'll be shown in OpenShift UI for `serviceAccount` field.
Suggestion: `Defines the service account`

File: api/v1/runtimecomponent_types.go [hunk 2]
```diff
@@ -84,6 +84,10 @@ type RuntimeComponentSpec struct {
 	// +operator-sdk:csv:customresourcedefinitions:order=12,type=spec,displayName="Probes"
 	Probes *RuntimeComponentProbes `json:"probes,omitempty"`
 
+	// TODO fix the ordering
```
Comment (leochr) [inline api/v1/runtimecomponent_types.go:87]: ServiceAccountName field's description (in API) should be marked as deprecated. `.spec.serviceAccount.name` overrides this if both are specified.

File: utils/utils.go [hunk 3]
```diff
@@ -809,6 +809,17 @@ func CustomizeServiceAccount(sa *corev1.ServiceAccount, ba common.BaseComponent,
...
+	mount := true
+	basa := ba.GetServiceAccount()
+	if basa != nil {
+		if basa.GetMountSAToken() != nil && !*basa.GetMountSAToken() {
+			// not nil and set to false
+			mount = false
+		}
+	}
+	sa.AutomountServiceAccountToken = &mount
+
```
Comment (leochr) [inline utils/utils.go:822]: We should set `automountServiceAccountToken: false` on Pod as well. This way, user can disable mount even for the non-default SA.

**Review by leochr — COMMENTED** [review]
@idlewis Thanks for the updates. Added some comments. 

We should add some tests as well. Fine to do that separately after this PR

File: utils/utils.go (hunk 3 above)
Comment (idlewis) [reply utils/utils.go:822]: @leochr 
Can I clarify this?
Is the suggestion that we will have a single config item (.spec.ServiceAccount.MountToken), and if that is set to false, then we will set both 
 * AutomountServiceAccountToken=false on the service account we create
 * automountServiceAccountToken=false on the deployment pod spec
 
Alternatively we could just set this _only_ on the pod spec. That should cover both the case of using our service account or a user specified one, and would make the code simpler.

Comment (leochr) [reply utils/utils.go:822]: @idlewis It's fine to set it only on the pod spec (as it takes precedence over SA)

**Comment by idlewis** [comment]
@leochr This is now ready for review

File: api/v1/runtimecomponent_types.go [hunk 4]
```diff
@@ -54,10 +54,15 @@ type RuntimeComponentSpec struct {
 	// +operator-sdk:csv:customresourcedefinitions:order=5,type=spec,displayName="Pull Secret",xDescriptors="urn:alm:descriptor:io.kubernetes:Secret"
 	PullSecret *string `json:"pullSecret,omitempty"`
 
+	// Deprecated. .spec.serviceAccount.name should be used instead. If both are specified, .spec.serviceAccount.name will override this.
 	// Name of the service account to use for deploying the application. A service account is automatically created if it's not specified.
```
Comment (leochr) [inline api/v1/runtimecomponent_types.go:58]: The second line of the description should be removed

File: api/v1/runtimecomponent_types.go [hunk 5]
```diff
@@ -54,10 +54,15 @@ type RuntimeComponentSpec struct {
...
 	// +operator-sdk:csv:customresourcedefinitions:order=6,type=spec,displayName="Service Account Name",xDescriptors="urn:alm:descriptor:com.tectonic.ui:text"
```
Comment (leochr) [inline api/v1/runtimecomponent_types.go:59]: I think it'll be good to hide this deprecated field from the OpenShift UI

File: api/v1/runtimecomponent_types.go [hunk 6]
```diff
@@ -54,10 +54,15 @@ type RuntimeComponentSpec struct {
...
 	ServiceAccountName *string `json:"serviceAccountName,omitempty"`
 
+	// The service account to use for deploying the application. A service account is automatically created if this is not specifed
```
Comment (leochr) [inline api/v1/runtimecomponent_types.go:62]: Need `.` at the end to be consistent

File: api/v1/runtimecomponent_types.go [hunk 7]
```diff
@@ -144,6 +149,17 @@ type RuntimeComponentSpec struct {
 	SecurityContext *corev1.SecurityContext `json:"securityContext,omitempty"`
 }
 
+// Defines the service account
+type RuntimeComponentServiceAccount struct {
+	// Whether the Service Account token should be mounted into the application pods. Default true
```
Comment (leochr) [inline api/v1/runtimecomponent_types.go:154]: To be consistent with some of the other default booleans:
`Default true` --> `Defaults to true.`

File: api/v1/runtimecomponent_types.go [hunk 8]
```diff
@@ -144,6 +149,17 @@ type RuntimeComponentSpec struct {
...
+	// +operator-sdk:csv:customresourcedefinitions:order=1,type=spec,displayName="Auto mount Service Account Token",xDescriptors="urn:alm:descriptor:com.tectonic.ui:booleanSwitch"
```
Comment (leochr) [inline api/v1/runtimecomponent_types.go:155]: Auto mount Service Account Token --> Mount Service Account Token

**Review by leochr — COMMENTED** [review]
@idlewis Thanks for the updates. Looks good. Added some minor comments about descriptions and UI rendering.

**Comment by idlewis** [comment]
@leochr I've addressed all the review comment.

File: api/v1/runtimecomponent_types.go (hunk 7 above)
Comment (leochr) [reply api/v1/runtimecomponent_types.go]: The period `.` at the end is needed

**Review by leochr — APPROVED** [review]

**Comment by leochr** [comment]
@idlewis Thanks for the updates. Just a minor update to add a period is needed (see [comment](https://github.com/application-stacks/runtime-component-operator/pull/575#discussion_r1355808978)). Looks good otherwise. Added approval.

**Pull request merged**